tools/bigquery-hive-external-table-loader/src/main/scala/com/google/cloud/bqhiveloader/SparkJobs.scala

```diff
@@ -0,0 +1,399 @@
+      .setMaxRetryDelay(Duration.ofSeconds(300))
+      .setInitialRpcTimeout(Duration.ofHours(8))
+      .setRpcTimeoutMultiplier(1.0d)
+      .setMaxRpcTimeout(Duration.ofHours(8))
+      .setJittered(false)
+      .build()
+
+    val bigqueryCreate: BigQuery = BigQueryOptions.newBuilder()
+      .setLocation(c.bqLocation)
+      .setCredentials(bqCreateCredentials)
+      .setProjectId(c.bqProject)
+      .setHeaderProvider(FixedHeaderProvider.create("user-agent", "BQHiveLoader 0.1"))
+      .setRetrySettings(retrySettings)
+      .build()
+      .getService
+
+    val bigqueryWrite: BigQuery = BigQueryOptions.newBuilder()
+      .setLocation(c.bqLocation)
+      .setCredentials(bqWriteCredentials)
+      .setProjectId(c.bqProject)
+      .setHeaderProvider(FixedHeaderProvider.create("user-agent", "BQHiveLoader 0.1"))
+      .setRetrySettings(retrySettings)
+      .build()
+      .getService
+
+    val gcs: Storage = StorageOptions.newBuilder()
+      .setCredentials(storageCreds)
+      .setProjectId(c.bqProject)
+      .setHeaderProvider(FixedHeaderProvider.create("user-agent", "BQHiveLoader 0.1"))
+      .setRetrySettings(retrySettings)
+      .build()
+      .getService
+
+    // TODO detect from metadata
+    val storageFormat = c.hiveStorageFormat
+        .map(ExternalTableManager.parseStorageFormat)
+        .getOrElse(Orc)
+
+    val filteredSchema = StructType(table.schema
+      .filterNot{x => c.dropColumns.contains(x.name)}
+      .filter{x => c.keepColumns.contains(x.name) || c.keepColumns.isEmpty}
+      .map{x =>
+        c.renameColumns.find(_._1 == x.name) match {
+          case Some((_, newName)) =>
+            x.copy(name = newName)
+          case _ =>
+            x
+        }
+      })
+
+    /* Create BigQuery table to be loaded */
+    NativeTableManager.createTableIfNotExists(c.bqProject, c.bqDataset, c.bqTable, c, filteredSchema, bigqueryWrite)
+
+    val externalTables: Seq[(Partition,TableInfo, Boolean)] = partitions.map{part =>
+      val extTable = createExternalTable(c.bqProject, c.tempDataset, c.bqTable,
+        table, part, storageFormat, bigqueryCreate, gcs, c.dryRun)
+
+      val renameOrcCols = if (!c.dryRun) {
+        val creation = waitForCreation(extTable.getTableId, timeoutMillis = 120000L, bigqueryCreate)
+        hasOrcPositionalColNames(creation)
+      } else true
+
+      (part, extTable, renameOrcCols)
+    }
+
+    val sql: Seq[String] = externalTables.map{x =>
+      val partition = x._1
+      val extTableId = x._2.getTableId
+      val schema = table.schema
+      val renameOrcCols = x._3
+      SQLGenerator.generateSelectFromExternalTable(
+        extTable = extTableId,
+        schema = schema,
+        partition = partition,
+        unusedColumnName = c.unusedColumnName,
+        formats = c.partColFormats.toMap,
+        renameOrcCols = renameOrcCols,
+        dropColumns = c.dropColumns,
+        keepColumns = c.keepColumns)
+    }
+    val unionSQL = sql.mkString("\nUNION ALL\n")
```

Comment:
it will be helpful to comment about the UNION ALL usage

Reply:
added comments